concordium-std/src/prims.rs

```diff
@@ -207,6 +207,51 @@ extern "C" {
     /// Slot time (in milliseconds) from chain meta data
     pub fn get_slot_time() -> u64;
 
+    // Cryptographic primitives
+
+    /// Verify an ed25519 signature. The public key is expected to be 32 bytes,
+    /// the signature is expected to be 64 bytes, and the message may be
+    /// variable length.
+    ///
+    /// The return value is 0 if verification fails, and 1 if it succeeds. No
+    /// other return values are possible.
+    pub fn verify_ed25519_signature(
+        public_key: *const u8,
+        signature: *const u8,
+        message: *const u8,
+        message_len: u32,
+    ) -> i32;
+
+    /// Verify an ed25519 signature. The public key is expected to be 33 bytes,
+    /// the signature is expected to be 64 bytes (serialized in compressed
+    /// format), and the message must be 32 bytes. For the signature to not leak
+    /// the key the message must be hashed with a cryptographically secure
+    /// hash before signing. Thus we only allow checking signatures on
+    /// 32-byte messages.
```

Comment:
Do we use the bitcoin-core implementation? According to https://github.com/bitcoin-core/secp256k1/blob/master/examples/ecdsa.c, the verify function assumes that you pre-hash the message.
Could we rename `message` to `message_hash`?

Reply:
I got that from the documentation. I can remove the reason, which is not that useful.

But the implementation we have only supports signing and checking signatures on hashes. So there is something to this.

Reply:
We are using the bitcoin-core implementation. Yes, I'll rename to message_hash.

Reply:
https://twitter.com/pwuille/status/1063582706288586752

Reply:
Which is linked from the library documentation https://docs.rs/secp256k1/latest/secp256k1/struct.Message.html

Reply:
Note that I do not understand the details, I just included what I saw in the underlying implementation documentation. Let me know if something needs to be changed.

Reply:
I found this leaking the key part fishy and I don't see how this could happen, but it is true that one should only sign hashes because otherwise one can forge signatures. I'm fine with removing the reason and renaming message to hash.

If we want to protect users from misusing this, we could enforce hashing inside the function. That's of course less flexible and I'm not saying we should do this, but it is a consideration.

Reply:
Makes sense.

Note that this is just verification. I will keep the API exposed by the underlying high-quality libraries that we use.